### ISSUE TITLE
Generated register function sequence should be sorted by the dependency tree

### ISSUE THREAD

**Issue by rockyprabowo** [Milad-Akarie/injectable#4]
On the project I'm currently working on, the generated file yields an unexpected result. The dependency (UserRepository) of another class (AuthenticationBloc) is registered much later after the registration of the dependent class (AuthenticationBloc), causing runtime issues during the instantiation of AuthenticationBloc singleton. 

This is the generated file with runtime issues:
```
// GENERATED CODE - DO NOT MODIFY BY HAND

// **************************************************************************
// InjectableConfigGenerator
// **************************************************************************

import 'package:simak_ux/app/helpers/random_generator.dart';
import 'package:simak_ux/app/delegates/simple_bloc_delegate.dart';
import 'package:simak_ux/blocs/theme/theme_bloc.dart';
import 'package:simak_ux/blocs/authentication/authentication_bloc.dart';
import 'package:simak_ux/repositories/user_repository.dart';
import 'package:simak_ux/blocs/login/login_bloc.dart';
import 'package:simak_ux/repositories/study_plan_repository.dart';
import 'package:simak_ux/repositories/study_result_repository.dart';
import 'package:simak_ux/repositories/user_profile_repository.dart';
import 'package:get_it/get_it.dart';

final getIt = GetIt.instance;
void $initGetIt({String environment}) {
  getIt
    ..registerFactory<RandomGenerator>(() => RandomGenerator())
    ..registerFactory<SimpleBlocDelegate>(() => SimpleBlocDelegate())
    ..registerSingleton<ThemeBloc>(ThemeBloc())
    ..registerSingleton<AuthenticationBloc>(
        AuthenticationBloc(getIt<UserRepository>()))
    ..registerSingleton<LoginBloc>(
        LoginBloc(getIt<UserRepository>(), getIt<AuthenticationBloc>()))
    ..registerFactory<StudyPlanRepository>(() => StudyPlanRepository())
    ..registerSingleton<StudyResultRepository>(StudyResultRepository())
    ..registerSingleton<UserRepository>(UserRepository()) // Causes runtime error, should be registered before AuthenticationBloc
    ..registerSingleton<UserProfileRepository>(UserProfileRepository());
}
```

It can be fixed by moving the dependency registration before the dependant manually:
```
// GENERATED CODE - DO NOT MODIFY BY HAND

// **************************************************************************
// InjectableConfigGenerator
// **************************************************************************

import 'package:simak_ux/app/helpers/random_generator.dart';
import 'package:simak_ux/app/delegates/simple_bloc_delegate.dart';
import 'package:simak_ux/blocs/theme/theme_bloc.dart';
import 'package:simak_ux/blocs/authentication/authentication_bloc.dart';
import 'package:simak_ux/repositories/user_repository.dart';
import 'package:simak_ux/blocs/login/login_bloc.dart';
import 'package:simak_ux/repositories/study_plan_repository.dart';
import 'package:simak_ux/repositories/study_result_repository.dart';
import 'package:simak_ux/repositories/user_profile_repository.dart';
import 'package:get_it/get_it.dart';

final getIt = GetIt.instance;
void $initGetIt({String environment}) {
  getIt
    ..registerFactory<RandomGenerator>(() => RandomGenerator())
    ..registerFactory<SimpleBlocDelegate>(() => SimpleBlocDelegate())
    ..registerSingleton<ThemeBloc>(ThemeBloc())
    ..registerSingleton<UserRepository>(UserRepository()) // UserRepository is registered before AuthenticationBloc
    ..registerSingleton<AuthenticationBloc>(
        AuthenticationBloc(getIt<UserRepository>()))
    ..registerSingleton<LoginBloc>(
        LoginBloc(getIt<UserRepository>(), getIt<AuthenticationBloc>()))
    ..registerFactory<StudyPlanRepository>(() => StudyPlanRepository())
    ..registerSingleton<StudyResultRepository>(StudyResultRepository())
    ..registerSingleton<UserProfileRepository>(UserProfileRepository());
}
```
The generator should be able to figure this out and sort the registration sequence correctly.

**Comment by Milad-Akarie** [comment]
Hey @rockyprabowo, good point, only singletons should be ordered based on their dependents because they're registered as instances. I'll see about that. 
but I don't think initiating all of these instances at startup time is a good idea, you should consider registering them as lazySingletons (@lazySingelton) so they're only initiated the first time they're called. also, I can't help but notice the absurd amount of the singletons you have, you should defiantly consider using factories instead. 

Thanks for reporting

**Comment by rockyprabowo** [comment]
I am just playing around with the library to see how well it behaves by putting random annotations on some classes where it makes sense, and stumbled upon this issue. I know that everything singleton will cause a lot of problems in the long run, but I think by making authentication-related stuff as (lazy) singletons is quite reasonable.

Anyway, thank you for the tips. Ireally appreciate it. I hope this issue will get sorted. For now, I just need to register UserRepository as singleton manually before the generated registration function run.

**Comment by Milad-Akarie** [comment]
@rockyprabowo  fixed in v0.2.0.
Thanks for reporting.
